File: packages/goblins/testing/src/python/redis.py
"""Redis Testcontainer for Python integration tests."""

from typing import Optional
from testcontainers.redis import RedisContainer as TCRedisContainer
import redis.asyncio as redis


class RedisContainer:
    """Redis container for integration testing with redis-py."""

    def __init__(
        self,
        image: str = "redis:7-alpine",
        password: Optional[str] = None,
        port: int = 6379
    ):
        self.image = image
        self.password = password
        self.port = port
        self._container: Optional[TCRedisContainer] = None
        self._client: Optional[redis.Redis] = None

    async def start(self) -> "RedisContainer":
        """Start the Redis container."""
        self._container = TCRedisContainer(image=self.image, port=self.port)

        if self.password:
            self._container = self._container.with_command(
                f"redis-server --requirepass {self.password}"
            )

        self._container.start()
        return self

    async def stop(self) -> None:
        """Stop the Redis container."""
        if self._client:
            await self._client.close()
            self._client = None

        if self._container:
            self._container.stop()
            self._container = None

    async def get_client(self) -> redis.Redis:
        """Get a redis-py async client."""
        if not self._client:
            if not self._container:
                raise RuntimeError("Container not started. Call start() first.")

            client_kwargs = {
                "host": self.get_host(),
                "port": self.get_port(),
                "decode_responses": True
            }

            if self.password:
                client_kwargs["password"] = self.password

            self._client = redis.Redis(**client_kwargs)

        return self._client

    def get_connection_string(self) -> str:
        """Get the Redis connection string."""
        auth = f":{self.password}@" if self.password else ""
        return f"redis://{auth}{self.get_host()}:{self.get_port()}"

    def get_host(self) -> str:
        """Get the container host."""
        if not self._container:
            raise RuntimeError("Container not started")
        return self._container.get_container_host_ip()

    def get_port(self) -> int:
        """Get the mapped port."""
        if not self._container:
            raise RuntimeError("Container not started")
        return int(self._container.get_exposed_port(self.port))

    async def __aenter__(self) -> "RedisContainer":
        """Async context manager entry."""
        return await self.start()

    async def __aexit__(self, exc_type, exc_val, exc_tb) -> None:
        """Async context manager exit."""
        await self.stop()

File: packages/goblins/testing/src/node/redis.ts
// Avoid importing types from `redis` during triage; use a permissive alias.
type RedisClientType = any;
import {
	GenericContainer,
	type StartedTestContainer,
	Wait,
} from "testcontainers";

export interface RedisOptions {
	/** Docker image to use (default: redis:7-alpine) */
	image?: string;
	/** Password for Redis */
	password?: string;
	/** Port to expose (0 = random) */
	port?: number;
}

export class RedisTestcontainer {
	private container: StartedTestContainer;
	private options: Required<Omit<RedisOptions, "password" | "port">> & {
		password?: string;
		port?: number;
	};
	private _client?: RedisClientType;

	private constructor(
		container: StartedTestContainer,
		options: Required<Omit<RedisOptions, "password" | "port">> & {
			password?: string;
			port?: number;
		},
	) {
		this.container = container;
		this.options = options;
	}

	static async start(options: RedisOptions = {}): Promise<RedisTestcontainer> {
		const opts = {
			image: options.image ?? "redis:7-alpine",
			password: options.password,
			port: options.port,
		};

		const env: Record<string, string> = {};
		let command: string[] | undefined;

		if (opts.password) {
			command = ["redis-server", "--requirepass", opts.password];
		}

		const containerBuilder = new GenericContainer(opts.image)
			.withEnvironment(env)
			.withExposedPorts(opts.port ?? 6379)
			.withWaitStrategy(Wait.forLogMessage(/Ready to accept connections/));

		if (command) {
			containerBuilder.withCommand(command);
		}

		const container = await containerBuilder.start();

		return new RedisTestcontainer(container, opts);
	}

	async stop(): Promise<void> {
		if (this._client) {
			await this._client.quit();
			this._client = undefined;
		}
		await this.container.stop();
	}

	async getClient(): Promise<RedisClientType> {
		if (!this._client) {
			// Lazy import to avoid dependency when not used
			const redis = await import("redis");
			const clientOptions: Record<string, unknown> = {
				socket: {
					host: this.getHost(),
					port: this.getPort(),
				},
			};

			if (this.options.password) {
				clientOptions.password = this.options.password;
			}

			this._client = redis.createClient(clientOptions) as RedisClientType;
			await this._client.connect();
		}
		return this._client;
	}

	getConnectionString(): string {
		const auth = this.options.password ? `:${this.options.password}@` : "";
		return `redis://${auth}${this.getHost()}:${this.getPort()}`;
	}

	getHost(): string {
		return this.container.getHost();
	}

	getPort(): number {
		return this.container.getMappedPort(this.options.port ?? 6379);
	}
}
